clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -592,9 +593,13 @@ public void close(final Timer timer) {
     void invokeCompletedOffsetCommitCallbacks() {
         while (true) {
             OffsetCommitCompletion completion = completedOffsetCommits.poll();
-            if (completion == null)
+            if (completion == null) {
                 break;
-            completion.invoke();
+            } else if (completion.exception instanceof FencedInstanceIdException) {
+                throw (FencedInstanceIdException) completion.exception;
```

Comment:
Hmm.. Why are we throwing this instead of invoking the callback?

Reply:
The goal here is to fail immediately, because we have detected a fenced exception in commit response queue, there should be no point retrying.

Reply:
Hmm one sec, a not-very-common-but-possible-pattern may be: 

1) user first call `commitAsync`, the response sent back has Fenced error; it would be kept.
2) user second call `commitAsync`, we would poll the completion object and throw immediately.
3) however, let's say user swallowed it and call `commitAsync` again, in this case we would proceed since the previous completion has been polled and hence there's no fence error any more, right?

I think the contract should be, once the consumer falls into the `fenced` state, it should always be in that state and hence always throw immediately for any following function calls. In this sense, we should probably change this logic a bit.

Reply:
> I think the contract should be, once the consumer falls into the fenced state, it should always be in that state and hence always throw immediately for any following function calls. In this sense, we should probably change this logic a bit.

Yes, that is what I was trying to get to above. Adding a `FENCED` state to `MemberState` would be a nice way to achieve this.